### PR TITLE
Feature/update from content

### PR DESCRIPTION
**Pull request by thives**
This simplifies applying updates from a remote source. 

Most of the time the updated executable comes from an outside source, in my case a database. 
It's more efficient to be able to call `self_replace_with(content)` instead of writing the new file content to a random temporary file on the filesystem and then call `self_replace(tmp_filename)` which  then copies the file to yet another temporary.

I took the liberty to refactor a bit to avoid code duplication due to my additions.

### CODE REVIEW

**Comment by mitsuhiko** [comment]
I’m surprised that you have a case where this is more efficient. It requires buffering in memory whereas writing directly to the disk and to then move has the advantage that you can stream bytes in.

I don’t feel compelled to add a way to dump bytes in directly for the most part. Can you share more about why you have that need?

**Comment by thives** [comment]
Good point, there isn't much benefit other than being more ergonomic. We're still having to store 3 executables during this process. It is good that you push back on this because my attempt here wouldn't be enough for my use case.

My use case is OTA on embedded (Linux) with limited resources (including NAND flash that we should try to not wear out unnecessarily). It should also handle any failures such as sporadic resets without human intervention, so the overall update process needs to be idempotent. I believe writing directly to a file with a unique and persistent name would be a part of a solution towards that. Ideally we want to be able to successfully update as many devices as possible in the shortest time possible using as little data traffic as possible (because of LTE data costs). To that end I also want to use a signed hashtree to both validate the new executable and to enable continuing a partial update process (outside of this crate) - increasing the success rate and lowering the risk of wearing out flaky devices.

In summary our focus is, in order:
- Low resource usage
- Robustness
- Autonomy

Here are some suggestions:
- How about replacing the &[u8] argument with a closure providing a BufWriter to the created temporary file? That would give us a 2 executables solution while also limiting the required buffer size. It also makes it easy to build a more sophisticated stack on top. 
- Add possibility to provide a name (not necessarily path) of the temporary file. 
- Add possibility to choose whether to overwrite an existing temporary file or to append to it.

**Comment by mitsuhiko** [comment]
So far I have seen the process of managing the temporary file outside the scope of this project. Here is for instance how Rye does it: https://github.com/astral-sh/rye/blob/ab8d5b433d5c4342c2bb125583c6bff4d29f5fbc/rye/src/cli/rye.rs#L303-L318

While it's possible to move some of this logic into this crate, there are so many subtle problems with tempfiles that I'm not sure if it's a great idea to try to have a one-size fits all kind of solution.

I would propose you try to make it work with `self_replace` today and if you learn something from your experience, we can look if we move some of the code you write around into this crate.
